File: DarKSward-main/server_log.py
#!/usr/bin/env python3
import http.server
import socketserver
import json
import urllib.parse
import asyncio
import websockets
import threading
from datetime import datetime
from http.server import HTTPServer

PORT = 8000
WS_PORT = 8001
LOG_FILE = "/tmp/worker_log.txt"

connected_clients = set()
log_lock = threading.Lock()


def log_print(msg, is_error=False):
    timestamp = datetime.now().strftime("%H:%M:%S")
    log_entry = f"[{timestamp}] {msg}"
    prefix = "\033[91m" if is_error else ""
    suffix = "\033[0m" if is_error else ""
    print(f"{prefix}{log_entry}{suffix}")
    with open(LOG_FILE, "a") as f:
        f.write(log_entry + "\n")


class MyHTTPRequestHandler(http.server.SimpleHTTPRequestHandler):
    def do_GET(self):
        log_print(f"{self.client_address[0]} GET {self.path}")

        if self.path.startswith("/log.html"):
            parsed = urllib.parse.urlparse(self.path)
            params = urllib.parse.parse_qs(parsed.query)
            text = params.get("text", ["(no text)"])[0]
            log_print(f"  WORKER: {text}")
            self.send_response(200)
            self.send_header("Content-type", "text/html")
            self.end_headers()
            self.wfile.write(b"OK")
            return

        if self.path.startswith("/error"):
            parsed = urllib.parse.urlparse(self.path)
            params = urllib.parse.parse_qs(parsed.query)
            msg = params.get("msg", ["(no msg)"])[0]
            log_print(f"  ERROR: {msg}", is_error=True)
            self.send_response(200)
            self.send_header("Content-type", "text/html")
            self.end_headers()
            self.wfile.write(b"OK")
            return

        if self.path.startswith("/ws"):
            log_print("WebSocket upgrade request received")
            self.send_response(400)
            self.end_headers()
            return

        return super().do_GET()

    def do_POST(self):
        content_length = int(self.headers.get("Content-Length", 0))
        body = self.rfile.read(content_length).decode("utf-8", errors="ignore")
        log_print(f"{self.client_address[0]} POST {self.path} - {body[:200]}")
        self.send_response(200)
        self.send_header("Content-type", "text/html")
        self.end_headers()
        self.wfile.write(b"OK")

    def log_message(self, format, *args):
        pass


async def websocket_handler(websocket, path):
    log_print(f"WebSocket client connected from {websocket.remote_address}")
    connected_clients.add(websocket)
    try:
        async for message in websocket:
            try:
                data = json.loads(message)
                msg_type = data.get("type", "unknown")
                if msg_type == "log":
                    tag = data.get("tag", "")
                    msg_text = data.get("message", "")
                    log_print(f"  [{tag}] {msg_text}")
                elif msg_type == "stage":
                    stage = data.get("stage", "")
                    status = data.get("status", "")
                    percent = data.get("percent", 0)
                    is_error = data.get("isError", False)
                    log_print(
                        f"  STAGE [{stage}] {percent}%: {status}"
                        + (" [ERROR]" if is_error else "")
                    )
                elif msg_type == "error":
                    log_print(f"  ERROR: {data.get('message', '')}", is_error=True)
                elif msg_type == "complete":
                    log_print("  COMPLETE")
                elif msg_type == "exploit_result":
                    log_print(
                        f"  ★★★ EXPLOIT RESULT: Device={data.get('deviceModel', 'Unknown')} jscBase={data.get('jscBase', 'N/A')} info={data.get('info', 'N/A')} ★★★"
                    )
                else:
                    log_print(f"  WS: {msg_type}")
            except json.JSONDecodeError:
                log_print(f"  WS RAW: {message}")
    except websockets.exceptions.ConnectionClosed:
        log_print("WebSocket client disconnected")
    finally:
        connected_clients.discard(websocket)


async def ws_main():
    async with websockets.serve(websocket_handler, "0.0.0.0", WS_PORT):
        log_print(f"WebSocket server running on ws://0.0.0.0:{WS_PORT}")
        await asyncio.Future()


def run_websocket_server():
    asyncio.run(ws_main())


if __name__ == "__main__":
    open(LOG_FILE, "w").close()

    ws_thread = threading.Thread(target=run_websocket_server, daemon=True)
    ws_thread.start()

    log_print(f"HTTP server running on http://0.0.0.0:{PORT}")
    log_print(f"WebSocket server running on ws://0.0.0.0:{WS_PORT}")
    log_print(f"Logs saved to {LOG_FILE}")
    log_print("-" * 50)

    with socketserver.TCPServer(("", PORT), MyHTTPRequestHandler) as httpd:
        httpd.serve_forever()

File: DarKSward-main/server_log.ts
import http from "http";
import fs from "fs";
import url from "url";
import path from "path";

const PORT = 8000;
const LOG_FILE = "/tmp/worker_log.txt";

function logPrint(msg: string, isError = false): void {
  const timestamp = new Date().toLocaleTimeString("zh-CN", { hour12: false });
  const logEntry = `[${timestamp}] ${msg}`;
  const prefix = isError ? "\x1b[91m" : "";
  const suffix = isError ? "\x1b[0m" : "";
  console.log(`${prefix}${logEntry}${suffix}`);
  fs.appendFileSync(LOG_FILE, logEntry + "\n");
}

const MIME_TYPES: Record<string, string> = {
  ".html": "text/html",
  ".js": "application/javascript",
  ".css": "text/css",
  ".json": "application/json",
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".gif": "image/gif",
  ".svg": "image/svg+xml",
  ".ico": "image/x-icon",
};

const httpServer = http.createServer((req, res) => {
  const clientIp = req.socket.remoteAddress || "unknown";
  const parsedUrl = url.parse(req.url || "", true);
  const host = req.headers.host || `localhost:${PORT}`;
  const serverHost = host.includes(':') ? host : `${host}:${PORT}`;

  if (req.method === "GET") {
    logPrint(`${clientIp} GET ${parsedUrl.pathname}`);

    if (parsedUrl.pathname === "/log") {
      const tag = (parsedUrl.query.tag as string) || "";
      const msg = (parsedUrl.query.msg as string) || "";
      if (tag === "WORKER") {
        logPrint(`  [WORKER] ${msg}`);
      } else if (tag) {
        logPrint(`  [${tag}] ${msg}`);
      } else {
        logPrint(`  ${msg}`);
      }
      res.writeHead(200, { "Content-Type": "text/html" });
      res.end("OK");
      return;
    }

    if (parsedUrl.pathname === "/log.html") {
      const text = (parsedUrl.query.text as string) || "(no text)";
      logPrint(`  WORKER: ${text}`);
      res.writeHead(200, { "Content-Type": "text/html" });
      res.end("OK");
      return;
    }

    if (parsedUrl.pathname === "/error") {
      const msg = (parsedUrl.query.msg as string) || "(no msg)";
      logPrint(`  ERROR: ${msg}`, true);
      res.writeHead(200, { "Content-Type": "text/html" });
      res.end("OK");
      return;
    }

    if (parsedUrl.pathname === "/ws") {
      logPrint("WebSocket upgrade request received");
      res.writeHead(400);
      res.end();
      return;
    }

    if (parsedUrl.pathname === "/stage") {
      const stage = (parsedUrl.query.stage as string) || "";
      const status = (parsedUrl.query.status as string) || "";
      const percent = parseInt(parsedUrl.query.percent as string) || 0;
      const isError = parsedUrl.query.isError === '1';
      logPrint(`  STAGE [${stage}] ${percent}%: ${status}${isError ? " [ERROR]" : ""}`);
      res.writeHead(200, { "Content-Type": "text/html" });
      res.end("OK");
      return;
    }

    if (parsedUrl.pathname === "/complete") {
      logPrint("  COMPLETE");
      res.writeHead(200, { "Content-Type": "text/html" });
      res.end("OK");
      return;
    }

    let filePath = parsedUrl.pathname === "/" ? "/index.html" : parsedUrl.pathname || "/index.html";
    const fullPath = path.join(process.cwd(), filePath);
    const ext = path.extname(fullPath);
    const mimeType = MIME_TYPES[ext] || "application/octet-stream";

    try {
      const content = fs.readFileSync(fullPath);
      res.writeHead(200, {
        "Content-Type": mimeType,
        "Cache-Control": "no-store, no-cache, must-revalidate",
        "Pragma": "no-cache",
        "Expires": "0",
      });
      res.end(content);
      return;
    } catch {
      res.writeHead(404);
      res.end("Not Found");
      return;
    }
  }

  if (req.method === "POST") {
    let body = "";
    req.on("data", (chunk) => (body += chunk));
    req.on("end", () => {
      logPrint(`${clientIp} POST ${parsedUrl.pathname} - ${body.slice(0, 200)}`);
      res.writeHead(200, { "Content-Type": "text/html" });
      res.end("OK");
    });
    return;
  }

  res.writeHead(404);
  res.end();
});

fs.writeFileSync(LOG_FILE, "");

logPrint(`HTTP server running on http://0.0.0.0:${PORT}`);
logPrint(`Logs saved to ${LOG_FILE}`);
logPrint("-".repeat(50));

httpServer.listen(PORT, "0.0.0.0", () => {
  logPrint(`Ready - request Host header determines server address`);
});